### PR TITLE
[Fixes #454] Log messages from HostingEngine are sometimes null 

### LINKED ISSUE #454: Log messages from HostingEngine are sometimes null
Example line of code causing the issue: https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs#L140

Suggestion (from @rynowak ): Change it to a lazy initialize pattern ex:
```
If (_cached == null)
{
  _cached = …;
}
```

Suggestion (from @loudejard): Add the missing piece
```
public override string ToString() => 
_cachedToString ?? Interlocked.CompareExchange(
                ref _cachedToString,
                $"Request starting {_httpContext.Request.Protocol} {_httpContext.Request.Method} {_httpContext.Request.Scheme}://{_httpContext.Request.Host}{_httpContext.Request.PathBase}{_httpContext.Request.Path}{_httpContext.Request.QueryString} {_httpContext.Request.ContentType} {_httpContext.Request.ContentLength}",
                null) ?? _cachedToString;
```

The `CompareExchange` will return `null` when it's successful, but we want it to return the new value (now stored in `_cachedToString`).

@loudej @lodejard

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak @loudej @lodejard @davidfowl 

I have tested this with the Diagnostics repo where the tests were failing and also on the other thread where we the scope was not being printed.

Also tested on Mono on Linux (however there seems to be an issue with Diagnostics repo on Linux which seems to be unrelated to my fix...will talk to @rowanmiller about it)

cc: @JunTaoLuo

### CODE REVIEW

File: src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs [hunk 1]
```diff
@@ -108,55 +107,75 @@ public HostingLogScope(HttpContext httpContext)
...
+                        new KeyValuePair<string, object>("Method", _request.Method),
+                        new KeyValuePair<string, object>("ContentType", _request.ContentType),
+                        new KeyValuePair<string, object>("ContentLength", _request.ContentLength),
+                        new KeyValuePair<string, object>("Scheme", _request.Scheme.ToString()),
+                        new KeyValuePair<string, object>("Host", _request.Host.ToString()),
+                        new KeyValuePair<string, object>("PathBase", _request.PathBase.ToString()),
+                        new KeyValuePair<string, object>("Path", _request.Path.ToString()),
+                        new KeyValuePair<string, object>("QueryString", _request.QueryString.ToString()),
+                    };
+                };
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs]: semicolon?

**Comment by rowanmiller** [comment]
@kichalla that test won't run on Linux because it uses a LocalDb instance (separate issue already filed to enable them on another database engine). I'll send a PR to skip it.

File: src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs [hunk 2]
```diff
@@ -108,55 +107,75 @@ public HostingLogScope(HttpContext httpContext)
...
+                return _cachedGetValues;
+            }
         }
 
         private class HostingRequestStarting : ILogValues
         {
             internal static readonly Func<object, Exception, string> Callback = (state, exception) => ((HostingRequestStarting)state).ToString();
 
-            private readonly HttpContext _httpContext;
+            private readonly HttpRequest _request;
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs]: This is nice, but why only in request starting? For example, we can do something similar for `_httpContext.Response` in `HostingRequestFinished`.

Comment (kichalla) [inline src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs]: The line for `ToString` was extremely long where as for others they are ok

File: src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs]: good catch...apparently the compiler seems ok with this :smile: ...will remove

**Comment by JunTaoLuo** [comment]
:shipit: from me. @pranavkm we are branching so this change will need to go into both release and dev pending approval?

**Comment by rynowak** [comment]
:shipit:

**Comment by kichalla** [comment]
2bb2f955327ea52c4b0536732d85b4a42f16e1a5

**Pull request closed without merging**